pymtl3/passes/sim/SimpleSchedulePass.py

```diff
@@ -113,35 +114,37 @@ def schedule_posedge_flip( self, top ):
     strs = []
     for x,y in hostobj_signals.items():
       if len(y) == 1:
-        strs.append( f"{repr(y[0])}._flip()" )
+        strs.append( f"    {repr(y[0])}._flip()" )
       elif x is top:
         for z in sorted(y, key=repr):
-          strs.append(f"{repr(z)}._flip()")
+          strs.append(f"    {repr(z)}._flip()")
       else:
-        pos = len(repr(x)) + 1
-        strs.append( f"x = {repr(x)}" )
+        repr_x = repr(x)
+        pos = len(repr_x) + 1
+        strs.append( f"    x = {repr_x}" )
 
         for z in sorted(y, key=repr):
-          strs.append(f"x.{repr(z)[pos:]}._flip()")
+          strs.append(f"    x.{repr(z)[pos:]}._flip()")
 
     if not strs:
       def no_double_buffer():
         pass
       top._sched.schedule_posedge_flip = [ no_double_buffer ]
 
     else:
-      src = """
-      def compile_double_buffer( s ):
-        def double_buffer():
-          {}
-        return double_buffer
-      """.format( "\n          ".join(strs) )
-
-      import py
-      # print(src)
+      lines = ['def compile_double_buffer( s ):'] + \
+              ['  def double_buffer():'] + \
+                strs + \
+              ['  return double_buffer']
+
+      # Shunning: The reason why we replace py.code.Source with exec(compile()) + linecache
+      # is because py.code.Source takes a full source code and divide them into
+      # a list of lines by newline character which scales very very poorly
+      # when the source code is huge. For some designs with 10K+ flip-flops
+      # the performance overhead becomes huge.
       l = locals()
-      custom_exec(py.code.Source( src ).compile(), globals(), l)
-
+      custom_exec( compile( '\n'.join(lines), filename='ff_flips', mode='exec' ), globals(), l)
```

Comment:
Is this the only place that uses `py.code`? If so can we remove the dependency on the `py` package?

Reply:
Did we list py as a dependency? I didn't find it in requirement.txt.